core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -1777,6 +1703,24 @@ private void checkDistressed() {
         throw new CairoError("Table '" + tableName + "' is distressed");
     }
 
+    private boolean checkScoreboardHasReadersBeforeLastCommittedTxn() {
```

Comment:
I wonder if this is a safe way to check for absent readers of older transactions. Let's consider the following situation:
```java
final long lastCommittedTxn = 2;

// Thread A (reader) - grabs read permit
txnScoreboard.acquireTxn(lastCommittedTxn);
// Thread B (reader) - lags behind and grabs read permit for the previous transaction
txnScoreboard.acquireTxn(lastCommittedTxn - 1);
// txnScoreboard.getMin() equals to 2 and will stay so

// Thread C (writer) - checkScoreboardHasReadersBeforeLastCommittedTxn
txnScoreboard.acquireTxn(lastCommittedTxn);
txnScoreboard.releaseTxn(lastCommittedTxn);
// The assertion fails while there are readers
Assert.assertTrue(txnScoreboard.getMin() != lastCommittedTxn);

// Thread A (reader) - releases read permit
txnScoreboard.releaseTxn(lastCommittedTxn);
// Thread B (reader) - releases read permit
txnScoreboard.releaseTxn(lastCommittedTxn - 1);
```

Am I missing something here?

Reply:
Should be fixed in https://github.com/questdb/questdb/pull/1778/commits/ee6968cf110ecf4e69455371d26b0503705da3e1
Solution is to not allow Out of Order transaction acquisition in the scoreboard. When OOO is detected `TableReader` has to re-read `_txn` file and retry with updated `txn` number

Reply:
I had to go for more significant changes in the scoreboard implementation at the end in 1bf83fe